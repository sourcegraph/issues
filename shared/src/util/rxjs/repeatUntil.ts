import { Observable } from 'rxjs'
import { repeatWhen, delay, takeWhile, repeat } from 'rxjs/operators'

/**
 * Mirrors values from the source observable and resubscribes to the source observable when it completes,
 * until an emission matches the provided condition. Resubscription is optionally delayed.
 */
export const repeatUntil = <T>(select: (value: T) => boolean, delayTime?: number) => (
    source: Observable<T>
): Observable<T> =>
    source.pipe(
        delayTime ? repeatWhen(completions => completions.pipe(delay(delayTime))) : repeat(),
        // Inclusive takeWhile so that the first value matching `select()` is emitted.
        takeWhile(value => !select(value), true)
    )
