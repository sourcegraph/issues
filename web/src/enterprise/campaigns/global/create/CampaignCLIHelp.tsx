import React, { useMemo } from 'react'
import { Collapsible } from '../../../../components/Collapsible'
import LanguageGoIcon from 'mdi-react/LanguageGoIcon'
import GithubCircleIcon from 'mdi-react/GithubCircleIcon'
import ExternalLinkIcon from 'mdi-react/ExternalLinkIcon'
import { highlightCode as _highlightCode } from '../../../../search/backend'
import { useObservable } from '../../../../../../shared/src/util/useObservable'
import { ThemeProps } from '../../../../../../shared/src/theme'
import LanguageTypescriptIcon from 'mdi-react/LanguageTypescriptIcon'
import { highlight } from 'highlight.js/lib/highlight'

interface Props extends ThemeProps {
    className?: string
    highlightCode?: typeof _highlightCode
}

const lsifAction = `{
  "scopeQuery": "repohasfile:go.mod repo:github -repohasfile:.github/workflows/lsif.yml",
  "steps": [
    {
      "type": "docker",
      "image": "add-lsif-to-build-pipeline-action"
    }
  ]
}`

const lsifGHAction = `name: LSIF
on:
  - push
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Generate LSIF data
        uses: sourcegraph/lsif-go-action@master
        with:
          verbose: "true"
      - name: Upload LSIF data
        uses: sourcegraph/lsif-upload-action@master
        with:
          github_token: \${{ secrets.GITHUB_TOKEN }}`

const lsifDockerfile = `FROM alpine:3
ADD ./github-action-workflow-golang.yml /tmp/workflows/

CMD mkdir -p .github/workflows && \\
  DEST=.github/workflows/lsif.yml; \\
  if [ ! -f .github/workflows/lsif.yml ]; then \\
    cp /tmp/workflows/github-action-workflow-golang.yml $DEST; \\
  else \\
    echo Doing nothing because existing LSIF workflow found at $DEST; \\
  fi`

const combyAction = `{
  "scopeQuery": "lang:go fmt.Sprintf",
  "steps": [
    {
      "type": "docker",
      "image": "comby/comby",
      "args": ["-in-place", "fmt.Sprintf(\\"%d\\", :[v])", "strconv.Itoa(:[v])", "-matcher", ".go", "-d", "/work"]
    },
    {
      "type": "docker",
      "image": "cytopia/goimports",
      "args": ["-w", "/work"]
    }
  ]
}`

const eslintAction = `{
  "scopeQuery": "repohasfile:yarn\\\\.lock repohasfile:tsconfig\\\\.json",
  "steps": [
    {
      "type": "docker",
      "image": "eslint-fix-action"
    }
  ]
}`

const eslintDockerfile = `FROM node:12-alpine3.10
CMD package_json_bkup=$(mktemp) && \\
  cp package.json $package_json_bkup && \\
  yarn -s --non-interactive --pure-lockfile --ignore-optional --ignore-scripts --ignore-engines --ignore-platform --no-progress && \\
  yarn add --ignore-workspace-root-test --non-interactive --ignore-optional --ignore-scripts --ignore-engines --ignore-platform --pure-lockfile -D @typescript-eslint/parser @typescript-eslint/eslint-plugin --no-progress eslint && \\
  node_modules/.bin/eslint \\
  --fix \\
  --plugin @typescript-eslint \\
  --parser @typescript-eslint/parser \\
  --parser-options '{"ecmaVersion": 8, "sourceType": "module", "project": "tsconfig.json"}' \\
  --rule '@typescript-eslint/prefer-optional-chain: 2' \\
  --rule '@typescript-eslint/no-unnecessary-type-assertion: 2' \\
  --rule '@typescript-eslint/no-unnecessary-type-arguments: 2' \
  --rule '@typescript-eslint/no-unnecessary-condition: 2' \\
  --rule '@typescript-eslint/no-unnecessary-type-arguments: 2' \\
  --rule '@typescript-eslint/prefer-includes: 2' \\
  --rule '@typescript-eslint/prefer-readonly: 2' \\
  --rule '@typescript-eslint/prefer-string-starts-ends-with: 2' \\
  --rule '@typescript-eslint/prefer-nullish-coalescing: 2' \\
  --rule '@typescript-eslint/no-non-null-assertion: 2' \\
  '**/*.ts'; \\
  mv $package_json_bkup package.json; \\
  rm -rf node_modules; \\
  yarn upgrade --latest --ignore-workspace-root-test --non-interactive --ignore-optional --ignore-scripts --ignore-engines --ignore-platform --no-progress typescript && \\
  rm -rf node_modules`

/**
 * A tutorial and a list of examples for campaigns using src CLI
 */
export const CampaignCLIHelp: React.FunctionComponent<Props> = ({
    className,
    isLightTheme,
    highlightCode = _highlightCode,
}) => {
    const formattedLsifAction = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: lsifAction,
                    isLightTheme,
                    fuzzyLanguage: 'json',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )
    const formattedLsifGHAction = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: lsifGHAction,
                    isLightTheme,
                    fuzzyLanguage: 'yaml',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )
    const formattedLsifDockerfile = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: lsifDockerfile,
                    isLightTheme,
                    fuzzyLanguage: 'Dockerfile',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )
    const formattedCombyAction = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: combyAction,
                    isLightTheme,
                    fuzzyLanguage: 'json',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )
    const formattedEslintAction = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: eslintAction,
                    isLightTheme,
                    fuzzyLanguage: 'json',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )
    const formattedEslintDockerfile = useObservable(
        useMemo(
            () =>
                highlightCode({
                    code: eslintDockerfile,
                    isLightTheme,
                    fuzzyLanguage: 'Dockerfile',
                    disableTimeout: false,
                }),
            [isLightTheme, highlightCode]
        )
    )

    const srcInstall = `# Configure your Sourcegraph instance:
$ export SRC_ENDPOINT=${window.location.protocol}//${window.location.host}

# Download the src binary for macOS:
$ curl -L $SRC_ENDPOINT/.api/src-cli/src_dawin_amd64 -o /usr/local/bin/src
# Download the src binary for Linux:
$ curl -L $SRC_ENDPOINT/.api/src-cli/src_linux_amd64 -o /usr/local/bin/src

# Set your personal access token:
$ export SRC_ACCESS_TOKEN=<YOUR TOKEN>
`

    return (
        <div className={className}>
            <h1>Create a campaign</h1>
            <div className="card">
                <div className="card-body p-3">
                    <h3>
                        1. Install the{' '}
                        <a href="https://github.com/sourcegraph/src-cli">
                            <code>src</code> CLI
                        </a>
                    </h3>
                    <div className="ml-2">
                        <p>Install and configure the src CLI:</p>
                        <pre className="ml-3">
                            <code
                                dangerouslySetInnerHTML={{
                                    __html: highlight('bash', srcInstall, true).value,
                                }}
                            />
                        </pre>
                        <p>
                            Make sure that <code>git</code> is installed and accessible by the src CLI.
                        </p>
                        <p>
                            To create and manage access tokens, click your username in the top right to open the user
                            menu, select <strong>Settings</strong>, and then <strong>Access tokens</strong>.
                        </p>
                    </div>
                    <h3>2. Create an action definition</h3>
                    <div className="ml-2 mb-1">
                        See below for examples and{' '}
                        <a
                            href="https://docs.sourcegraph.com/user/campaigns/creating_campaign_from_patches"
                            target="_blank"
                            rel="noopener noreferrer"
                        >
                            read the documentation on what actions can can do.
                        </a>
                        .
                    </div>
                    <h3>3. Optional: See repositories the action would run over</h3>
                    <div className="ml-2 mb-2">
                        <pre className="ml-3">
                            <code
                                dangerouslySetInnerHTML={{
                                    __html: highlight('bash', '$ src actions scope-query -f action.json', true).value,
                                }}
                            />
                        </pre>
                    </div>
                    <h3>4. Create a set of patches by executing the action over repositories</h3>
                    <div className="ml-2 mb-2">
                        <pre className="ml-3">
                            <code
                                dangerouslySetInnerHTML={{
                                    __html: highlight(
                                        'bash',
                                        '$ src actions exec -f action.json -create-patchset',
                                        true
                                    ).value,
                                }}
                            />
                        </pre>
                    </div>
                    <div className="alert alert-info mt-2">
                        <a
                            href=" https://docs.sourcegraph.com/user/campaigns/creating_campaign_from_patches"
                            rel="noopener noreferrer"
                            target="_blank"
                        >
                            Take a look at the documentation for more detailed steps and additional information.{' '}
                            <small>
                                <ExternalLinkIcon className="icon-inline" />
                            </small>
                        </a>
                    </div>
                </div>
            </div>
            <a id="examples" />
            <h2 className="mt-2">Examples</h2>
            <ul className="list-group mb-3">
                <li className="list-group-item p-2">
                    <Collapsible
                        title={
                            <h3 className="mb-0">
                                <GithubCircleIcon className="icon-inline ml-1 mr-2" /> Add a GitHub action to upload
                                LSIF data to Sourcegraph
                            </h3>
                        }
                        titleClassName="flex-grow-1"
                    >
                        <div>
                            <p>
                                Our goal for this campaign is to add a GitHub Action that generates and uploads LSIF
                                data to Sourcegraph by adding a .github/workflows/lsif.yml file to each repository that
                                doesn’t have it yet.
                            </p>
                            <p>
                                In order to build the Docker image, we first need to create a file called
                                <code>github-action-workflow-golang.yml</code> with the following content:
                            </p>

                            {formattedLsifGHAction && (
                                <code dangerouslySetInnerHTML={{ __html: formattedLsifGHAction }} />
                            )}
                            <p>And a Dockerfile:</p>
                            {formattedLsifDockerfile && (
                                <code dangerouslySetInnerHTML={{ __html: formattedLsifDockerfile }} />
                            )}
                            <p className="mt-2">Build the docker image:</p>
                            <p>
                                <code>docker build -t add-lsif-to-build-pipeline-action .</code>
                            </p>
                            {formattedLsifAction && <code dangerouslySetInnerHTML={{ __html: formattedLsifAction }} />}
                        </div>
                    </Collapsible>
                </li>
                <li className="list-group-item p-2">
                    <Collapsible
                        title={
                            <h3 className="mb-0">
                                <LanguageGoIcon className="icon-inline ml-1 mr-2" /> Refactor Go code using Comby
                            </h3>
                        }
                        titleClassName="flex-grow-1"
                    >
                        <div>
                            <p>
                                Our goal for this campaign is to simplify Go code by using Comby to rewrite calls to
                                <code>fmt.Sprintf("%d", arg)</code> with <code>strconv.Itoa(:[v])</code>. The semantics
                                are the same, but one more cleanly expresses the intention behind the code.
                            </p>
                            <p className="text-muted">
                                Note: Learn more about Comby and what it’s capable of at{' '}
                                <a href="https://comby.dev" target="_blank" rel="noopener noreferrer">
                                    comby.dev
                                </a>
                            </p>
                            <p>
                                To do that we use two Docker containers. One container launches Comby to rewrite the the
                                code in Go files and the other runs goimports to update the import statements in the
                                updated Go code so that strconv is correctly imported and, possibly, fmt is removed.
                            </p>
                            {formattedCombyAction && (
                                <code dangerouslySetInnerHTML={{ __html: formattedCombyAction }} />
                            )}
                        </div>
                    </Collapsible>
                </li>
                <li className="list-group-item p-2">
                    <Collapsible
                        title={
                            <h3 className="mb-0">
                                <LanguageTypescriptIcon className="icon-inline ml-1 mr-2" /> Upgrade to a new TypeScript
                                version
                            </h3>
                        }
                        titleClassName="flex-grow-1"
                    >
                        <div>
                            <p>
                                Our goal for this campaign is to convert all TypeScript code synced to our Sourcegraph
                                instance to make use of new TypeScript features. To do this we convert the code, then
                                update the TypeScript version.
                            </p>
                            <p>
                                Our goal for this campaign is to convert all TypeScript code synced to ouTo convert the
                                code we install and run ESLint with the desired typescript-eslint rules, using the --fix
                                flag to automatically fix problems. We then update the TypeScript version using yarn
                                upgrade.
                            </p>
                            <p>
                                The first thing we need is a Docker container in which we can freely install and run
                                ESLint. Here is the Dockerfile:
                            </p>
                            {formattedEslintDockerfile && (
                                <code dangerouslySetInnerHTML={{ __html: formattedEslintDockerfile }} />
                            )}
                            <p className="mt-2">
                                When turned into an image and run as a container, the instructions in this Dockerfile
                                will do the following:
                            </p>
                            <ol>
                                <li>
                                    Copy the current package.json to a backup location so that we can install ESLint
                                    without changes to the original package.json
                                </li>
                                <li>Install all dependencies &amp; add eslint with the typescript-eslint plugin</li>
                                <li>
                                    Run eslint --fix with a set of TypeScript rules to detect and fix problems over all
                                    *.ts files
                                </li>
                                <li>Restore the original package.json from its backup location</li>
                                <li>Run yarn upgrade to update the typescript version</li>
                            </ol>
                            <p>
                                Before we can run it as an action we need to turn it into a Docker image, by running the
                                following command in the directory where the Dockerfile was saved:
                            </p>
                            <code>docker build -t eslint-fix-action .</code>
                            <p>That builds a Docker image and names it eslint-fix-action.</p>
                            <p>Once that is done we’re ready to define our action:</p>
                            {formattedEslintAction && (
                                <code dangerouslySetInnerHTML={{ __html: formattedEslintAction }} />
                            )}
                        </div>
                    </Collapsible>
                </li>
            </ul>
        </div>
    )
}
