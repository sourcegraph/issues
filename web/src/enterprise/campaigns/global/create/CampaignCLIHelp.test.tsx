import React from 'react'
import renderer from 'react-test-renderer'
import { CampaignCLIHelp } from './CampaignCLIHelp'
import { of } from 'rxjs'

describe('CampaignCLIHelp', () => {
    test('renders', () => {
        const result = renderer.create(<CampaignCLIHelp isLightTheme={true} highlightCode={({ code }) => of(code)} />)
        expect(result.toJSON()).toMatchSnapshot()
    })
})
