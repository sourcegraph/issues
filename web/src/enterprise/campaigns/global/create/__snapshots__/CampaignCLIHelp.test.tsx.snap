// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignCLIHelp renders 1`] = `
<div>
  <h1>
    Create a campaign
  </h1>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <p
        className="alert alert-info"
      >
        Follow the step-by-step guide to get started with your first campaign. You can also find examples at the bottom of this page.
      </p>
      <h3>
        1. Install the src cli
      </h3>
      <div
        className="ml-2"
      >
        <p>
          If you have not already, first install and configure the src CLI to point to your Sourcegraph instance. This guide will get you the most recent compatible version with your Sourcegraph instance.
        </p>
        <h4>
          Configure the endpoint of your Sourcegraph instance
        </h4>
        <div>
          <code>
            export SRC_ENDPOINT=
            http:
            //
            localhost
          </code>
          <br />
          <p>
            <strong>
              Tip:
            </strong>
             You might want to put this in your shell config.
          </p>
        </div>
        <h4>
          Download the src CLI
        </h4>
        <h4>
          macOS
        </h4>
        <p>
          <code>
            curl -L $
            {
            SRC_ENDPOINT
            }
            /.api/src-cli/src_dawin_amd64 -o /usr/local/bin/src
          </code>
        </p>
        <h4>
          Linux
        </h4>
        <p>
          <code>
            curl -L $
            {
            SRC_ENDPOINT
            }
            /.api/src-cli/src_linux_amd64 -o /usr/local/bin/src
          </code>
        </p>
        <p>
          Also, make sure that git is installed and accessible by src.
        </p>
        <h4>
          Grating access to src CLI
        </h4>
        <p>
          To acquire the access token, visit your Sourcegraph instance, click your username in the top right to open the user menu, select Settings, and then select 
          <strong>
            Access tokens
          </strong>
           
          in the left hand menu. Then expose it to SRC_CLI using the following command:
        </p>
        <div>
          <code>
            export SRC_ACCESS_TOKEN=&lt;YOUR TOKEN&gt;
          </code>
          <br />
          <p>
            <strong>
              Tip:
            </strong>
             You might want to put this in your shell config.
          </p>
        </div>
      </div>
      <h3>
        2. Create an action JSON file (e.g. action.json) that contains an action definition
      </h3>
      <div
        className="ml-2 mb-1"
      >
        See below for examples of those files and
         
        <a
          href="https://docs.sourcegraph.com/user/campaigns#defining-an-action"
          rel="noopener noreferrer"
          target="_blank"
        >
          read the docs on what they can do
        </a>
        .
      </div>
      <h3>
        3. Optional: See repositories the action would run over
      </h3>
      <div
        className="ml-2 mb-2"
      >
        <code>
          src actions scope-query -f action.json
        </code>
      </div>
      <h3>
        4. Create a set of patches by executing the action over repositories
      </h3>
      <div
        className="ml-2 mb-2"
      >
        <code>
          src actions exec -f action.json &gt; patches.json
        </code>
      </div>
      <h3>
        5. Save the patches in Sourcegraph by creating a patch set
      </h3>
      <div
        className="ml-2 mb-2"
      >
        <code>
          src campaign patchset create-from-patches &lt; patches.json
        </code>
      </div>
      <h3>
        5. Create a campaign based on the patch set
      </h3>
      <div
        className="ml-2 mb-2"
      >
        <code>
          src campaigns create -branch=&lt;branch-name&gt; -patchset=&lt;patchset-ID-returned-by-previous-command&gt;
        </code>
      </div>
      <div
        className="ml-2"
      >
        <a
          href="https://docs.sourcegraph.com/user/campaigns#creating-a-campaign-using-the-src-cli"
          rel="noopener noreferrer"
          target="_blank"
        >
          Read on detailed steps and documentation in the docs.
           
          <small>
            <svg
              className="mdi-icon icon-inline"
              fill="currentColor"
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
              />
            </svg>
          </small>
        </a>
      </div>
    </div>
  </div>
  <a
    id="examples"
  />
  <h2
    className="mt-2"
  >
    Examples
  </h2>
  <ul
    className="list-group mb-3"
  >
    <li
      className="list-group-item p-2"
    >
      <div
        className=""
      >
        <div
          className="d-flex justify-content-between align-items-center position-relative"
        >
          <button
            aria-label="Expand section"
            className="d-flex btn btn-icon collapsible__expand-btn stretched-link"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-label="Expand section"
              className="mdi-icon icon-inline"
              fill="currentColor"
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
              />
            </svg>
          </button>
          <span
            className="flex-grow-1"
          >
            <h3
              className="mb-0"
            >
              <svg
                className="mdi-icon icon-inline ml-1 mr-2"
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M12,2A10,10 0 0,0 2,12C2,16.42 4.87,20.17 8.84,21.5C9.34,21.58 9.5,21.27 9.5,21C9.5,20.77 9.5,20.14 9.5,19.31C6.73,19.91 6.14,17.97 6.14,17.97C5.68,16.81 5.03,16.5 5.03,16.5C4.12,15.88 5.1,15.9 5.1,15.9C6.1,15.97 6.63,16.93 6.63,16.93C7.5,18.45 8.97,18 9.54,17.76C9.63,17.11 9.89,16.67 10.17,16.42C7.95,16.17 5.62,15.31 5.62,11.5C5.62,10.39 6,9.5 6.65,8.79C6.55,8.54 6.2,7.5 6.75,6.15C6.75,6.15 7.59,5.88 9.5,7.17C10.29,6.95 11.15,6.84 12,6.84C12.85,6.84 13.71,6.95 14.5,7.17C16.41,5.88 17.25,6.15 17.25,6.15C17.8,7.5 17.45,8.54 17.35,8.79C18,9.5 18.38,10.39 18.38,11.5C18.38,15.32 16.04,16.16 13.81,16.41C14.17,16.72 14.5,17.33 14.5,18.26C14.5,19.6 14.5,20.68 14.5,21C14.5,21.27 14.66,21.59 15.17,21.5C19.14,20.16 22,16.42 22,12A10,10 0 0,0 12,2Z"
                />
              </svg>
               Add a GitHub action to upload LSIF data to Sourcegraph
            </h3>
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item p-2"
    >
      <div
        className=""
      >
        <div
          className="d-flex justify-content-between align-items-center position-relative"
        >
          <button
            aria-label="Expand section"
            className="d-flex btn btn-icon collapsible__expand-btn stretched-link"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-label="Expand section"
              className="mdi-icon icon-inline"
              fill="currentColor"
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
              />
            </svg>
          </button>
          <span
            className="flex-grow-1"
          >
            <h3
              className="mb-0"
            >
              <svg
                className="mdi-icon icon-inline ml-1 mr-2"
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M2.64,10.33L2.62,10.27L2.84,10L2.96,9.92H6.8L6.83,10L6.65,10.26L6.54,10.32L2.64,10.33M1.03,11.31L1,11.26L1.22,10.97L1.34,10.91H6.24L6.29,11L6.21,11.24L6.11,11.31H1.03M3.63,12.3L3.59,12.24L3.75,11.96L3.85,11.9H6L6.07,11.97L6.05,12.22L5.97,12.3H3.63M14.78,10.14L13,10.61C12.81,10.65 12.8,10.66 12.66,10.5C12.5,10.32 12.39,10.21 12.16,10.1C11.5,9.76 10.83,9.86 10.22,10.25C9.5,10.73 9.11,11.42 9.12,12.3C9.13,13.16 9.72,13.87 10.57,14C11.3,14.09 11.91,13.83 12.4,13.28L12.69,12.89H10.62C10.4,12.89 10.35,12.75 10.42,12.57L10.97,11.39C11,11.33 11.08,11.22 11.24,11.22H14.68C14.83,10.72 15.09,10.26 15.43,9.81C16.21,8.78 17.16,8.24 18.43,8C19.5,7.82 20.56,7.93 21.5,8.57C22.34,9.15 22.87,9.93 23,10.96C23.19,12.41 22.76,13.59 21.76,14.61C21.05,15.33 20.18,15.78 19.19,16L18.33,16.08C17.35,16.06 16.46,15.78 15.71,15.13C15.19,14.68 14.83,14.14 14.65,13.5C14.5,13.74 14.38,13.97 14.21,14.2C13.44,15.22 12.43,15.85 11.15,16C10.1,16.16 9.12,15.95 8.26,15.31C7.47,14.71 7,13.91 6.9,12.92C6.76,11.75 7.1,10.7 7.81,9.78C8.57,8.78 9.58,8.15 10.82,7.92C11.82,7.74 12.79,7.86 13.66,8.44C14.23,8.82 14.63,9.34 14.9,9.96C14.94,10.05 14.9,10.11 14.78,10.14M20.89,11.74L20.86,11.38C20.67,10.32 19.69,9.72 18.67,9.95C17.66,10.17 17,10.8 16.79,11.81C16.6,12.65 17,13.5 17.77,13.84C18.36,14.1 18.96,14.06 19.53,13.78C20.37,13.35 20.84,12.66 20.89,11.74Z"
                />
              </svg>
               Refactor Go code using Comby
            </h3>
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item p-2"
    >
      <div
        className=""
      >
        <div
          className="d-flex justify-content-between align-items-center position-relative"
        >
          <button
            aria-label="Expand section"
            className="d-flex btn btn-icon collapsible__expand-btn stretched-link"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-label="Expand section"
              className="mdi-icon icon-inline"
              fill="currentColor"
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
              />
            </svg>
          </button>
          <span
            className="flex-grow-1"
          >
            <h3
              className="mb-0"
            >
              <svg
                className="mdi-icon icon-inline ml-1 mr-2"
                fill="currentColor"
                height={24}
                viewBox="0 0 24 24"
                width={24}
              >
                <path
                  d="M3,3H21V21H3V3M13.71,17.86C14.21,18.84 15.22,19.59 16.8,19.59C18.4,19.59 19.6,18.76 19.6,17.23C19.6,15.82 18.79,15.19 17.35,14.57L16.93,14.39C16.2,14.08 15.89,13.87 15.89,13.37C15.89,12.96 16.2,12.64 16.7,12.64C17.18,12.64 17.5,12.85 17.79,13.37L19.1,12.5C18.55,11.54 17.77,11.17 16.7,11.17C15.19,11.17 14.22,12.13 14.22,13.4C14.22,14.78 15.03,15.43 16.25,15.95L16.67,16.13C17.45,16.47 17.91,16.68 17.91,17.26C17.91,17.74 17.46,18.09 16.76,18.09C15.93,18.09 15.45,17.66 15.09,17.06L13.71,17.86M13,11.25H8V12.75H9.5V20H11.25V12.75H13V11.25Z"
                />
              </svg>
               Upgrade to a new TypeScript version
            </h3>
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
